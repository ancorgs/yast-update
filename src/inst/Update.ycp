S/**
 * Module:		Update.ycp
 *
 * Authors:		Anas Nashif (nashif@suse.de)
 *
 * Purpose:		Update module
 *
 *
 * $Id$
 */
{
    module "Update";

    global integer affected_count = 0;
    // number of packages affected by update

    global integer unknown_count = 0;
    // number of packages unknown by update (non-suse packages)

    global boolean kernel_affected = false;
    // if the kernel was affected in _normal_ update mode

    global boolean deleteOldPackages = false;
    // Flag is set true if the user decides to delete unmaintained packages

    global integer last_runlevel = -1;

    global string selectedRootPartition = "";
    // selected rootpartition for the update

    global map validRootPartitions = $[];
    // available rootpartitions (key) and SuSE versions (value)

    global boolean onlyUpdateSelected = false;
    // Only an update, NOT an upgrade

    global boolean updateEvaluated = false;
    // Flag if the updatelist has already evaluated

    global string installedName = "";
    // Name of the targetsystem

    global string installedVersion = "";
    // Version of the targetsystem

    global string updateVersion = "";
    // Version of the source medium

    global boolean updateBasePackages = false;
    // Flag, if the basesystem have to be installed

    global integer packagesInstalled = 0;
    // counter for installed packages

    /**
     * List of mounted partitions, activated swap partitions and loop devices.
     * Used for reversing action if mode is changed from update to new
     * installation or if root directory for update is changed. Each item is
     * list [string value, string type], where type is "mount", "swap", "loop",
     * value is path to mounted directory, swap partition name or loop device
     * name according to type value.
     */
    list activated = [];

    /*-----------------------------------------------------------------------
     * GLOBAL FUNCTIONS
     *-----------------------------------------------------------------------*/

    /**
     * Unmount all mounted partitions, deactivate swaps, detach loopback
     * devices. Uses list activated to make actions in reverse order.
     * @return void
     */
    global define void UnmountPartitions() ``{
	 y2milestone("Unmount partitions: %1", activated);

	 foreach(list info, activated, ``{
		string file = info[0]:"";
		string type = info[1]:"";

		if (file != "" && type != "")
		{
		    if (type == "mount")
		    {
			if (SCR::Execute (.target.umount, file) != 0)
			{
			    y2error("Cannot unmount partition %1", file);
			}
		    }
		    else if (type == "swap")
		    {
			if (SCR::Execute (.target.bash, "/sbin/swapoff " + file) != 0)
			{
			    y2error("Cannot deactivate swap %1", file);
			}
		    }
		    else if (type == "loop")
		    {
			if (WFM::Execute(.local.bash, "losetup -d " + file) != 0)
			{
			    y2error("Cannot deactivate loopback device %1", file);
			}
		    }
		}
	    }
	);

        SCR::Write(.lvm.deactivate, "");
	SCR::Write(.md.deactivate, "");

	// clear activated list
	activated = [];
    };

    /**
     * Add information about mounted partition to internal list.
     * @param partinfo partinfo has to be list with exactly two strings, see description of list "activated"
     * @return void
     */
    global define void AddMountedPartition(list partinfo) ``{
	y2debug("Added partition info: %1", partinfo);
	activated = prepend(activated, partinfo);
	y2debug("Total partition info: %1", activated);
    }

}


