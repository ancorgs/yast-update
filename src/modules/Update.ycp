/**
 * Module:		Update.ycp
 *
 * Authors:		Anas Nashif <nashif@suse.de>
 *			Arvin Schnell <arvin@suse.de>
 *
 * Purpose:		Update module
 *
 * $Id$
 */
{
    module "Update";

    import "Packages";
    import "ProductFeatures";
    import "Stage";

    // number of packages to install
    global integer packages_to_install = 0;

    // number of packages to update
    global integer packages_to_update = 0;

    // number of packages to remove
    global integer packages_to_remove = 0;

    // number of packages unknown (problematic) by update
    global integer unknown_packages = 0;

    // number of errors (packages?) returned by solver
    global integer solve_errors = 0;

    // Flag is set true if the user decides to delete unmaintained packages
    global boolean deleteOldPackages = ProductFeatures::delete_old_packages;

    // don't allow upgrade only update
    global boolean disallow_upgrade = false;

    global boolean did_init1 = false;

    global boolean did_init2 = false;


    global integer last_runlevel = -1;


    global boolean backup_modified = true;
    global boolean backup_sysconfig = true;
    global boolean remove_old_backups = false;
    global string backup_path = "/var/adm/backup";


    // Only an update, NOT an upgrade
    global boolean onlyUpdateInstalled = ProductFeatures::onlyUpdateInstalled;

    global string selected_selection = "";

    global boolean products_incompatible = false;


    /*
     *  Information about old and new product.
     *
     *  They do contain:
     *    name (string), e.g. "SuSE Linux"
     *    version (string), e.g. "9.1"
     *    nameandversion (string), e.g. "SuSE Linux 9.1"
     *
     *  They may contain:
     *    major (integer), e.g. "9"
     *    minor (integer), e.g. "1"
     *
     *  They are not initialized during the mounting of the filesystems so don't
     *  be so foolish to used them in RootPart.ycp.
     */

    // Version of the targetsystem
    global map <string, any> installedVersion = $[];

    // Version of the source medium
    global map <string, any> updateVersion = $[];


    // Flag, if the basesystem have to be installed
    global boolean updateBasePackages = false;

    // counter for installed packages
    global integer packagesInstalled = 0;


    // see bug #40358
    global boolean manual_interaction = false;

    // are the products (installed and to update) compatible?
    boolean _products_compatible = nil;




    /*-----------------------------------------------------------------------
     * GLOBAL FUNCTIONS
     *-----------------------------------------------------------------------*/


/**
 * Check if installed product and product to upgrade to are compatible
 * @return boolean true if update is possible
 */
global boolean ProductsCompatible () {
    if (_products_compatible == nil)
    {
	if (Stage::normal ())
	{
	    // check if name of one of the products on the installation
	    // media is same as one of the installed products
	    // assuming that multiple products on installation media
	    // are compatible and compatibility is transitive
	    list<map<string,string> > inst = (list<map<string,string> >)
		Pkg::TargetProducts ();
	    list<string> inst_names = maplist (map<string,string> p, inst, {
		return p["name"]:"";
	    });
	    list<string> to_install = maplist (integer src,
		Pkg::SourceGetCurrent (true),
	    {
		map<string,string> prod_info = Pkg::SourceProduct (src);
		return prod_info["name"]:"";
	    });
	    y2milestone ("Installed products: %1", inst_names);
	    y2milestone ("Products on installatino media: %1", to_install);

	    string equal_product = find (string i, inst_names, {
		string found = find (string u, to_install, {
		    return u == i;
		});
		return found != nil;
	    });
	    _products_compatible = equal_product != nil;
	}
	else
	{
	    _products_compatible = true; // FIXME this is temporary
	}
	y2milestone ("Products found compatible: %1", _products_compatible);
    }
    return _products_compatible;
}

global void IgnoreProductCompatibility () {
    _products_compatible = true;

}

    /**
     *
     */
    global define void Reset ()
    {
	deleteOldPackages = ProductFeatures::delete_old_packages;
	disallow_upgrade = false;

	manual_interaction = false;
	products_incompatible = false;
	_products_compatible = nil;

	backup_modified = true;
	backup_sysconfig = true;
	remove_old_backups = false;
	backup_path = "/var/adm/backup";
    }


    /**
     *
     */
    global define void fill_version_map (map <string, any>& data)
    {
	data["nameandversion"] = data["name"]:"?" + " " + data["version"]:"?";

	list <string> tmp1 = splitstring (data["version"]:"", ".");

	integer tmp2 = tointeger (tmp1[0]:"-1");
	if (tmp2 >= 0)
	    data["major"] = tmp2;

	integer tmp3 = tointeger (tmp1[1]:"-1");
	if (tmp3 >= 0)
	    data["minor"] = tmp3;
    }


    /**
     *
     */
    global define boolean GetProductName ()
    {
	installedVersion = $[];
	updateVersion = $[];

	/*
	 *  Be careful: The maps returned by from Pkg::TargetProducts and
	 *  Pkg::SourceProductData have different content.
	 */

	// get old product name

	list old_products = Pkg::TargetProducts ();
	map old_product = old_products[0]:$[];
	y2milestone ("First installed product %1", old_product);

	installedVersion["name"] = old_product["name"]:"?";
	installedVersion["version"] = old_product["version"]:"?";
	fill_version_map (installedVersion);

	// get new product name

	integer num = size (Packages::theSources);
	if (num <= 0)
	{
	    y2error ("No source");
	    updateVersion["name"] = "?";
	    updateVersion["version"] = "?";
	    fill_version_map (updateVersion);
	    return false;
	}

	map new_product = Pkg::SourceProductData (Packages::theSources[num-1]:0);
	y2milestone ("First source product %1", new_product);
	if (new_product == nil)
	{
	    y2error ("Screwed source");
	    updateVersion["name"] = "?";
	    updateVersion["version"] = "?";
	    fill_version_map (updateVersion);
	    return false;
	}

	updateVersion["name"] = new_product["productname"]:"?";
	updateVersion["version"] = new_product["productversion"]:"?";
	fill_version_map (updateVersion);

	y2milestone ("update from %1 to %2", installedVersion, updateVersion);

	return true;
    }


    /**
     * Get all available base selections sorted in reverse order
     * (highest ordered bases selection comes first).
     */
    global define list<string> GetBaseSelections ()
    {
	list<string> available_base_selections = sort (string x, string y, Pkg::GetSelections (`available, "base"), {
	    map xmap = Pkg::SelectionData(x);
	    map ymap = Pkg::SelectionData(y);
	    return (xmap["order"]:"" > ymap["order"]:"");
	});
	y2milestone ("available_base_selections %1", available_base_selections);
	return available_base_selections;
    }


    /**
     * Propose a selection for the update and save it's name in
     * Update::selected_selection.
     */
    global define void ProposeSelection ()
    {
	selected_selection = "";

	list available_selections = GetBaseSelections ();

	// first try: installed one

	list tmp1 = Pkg::GetSelections (`installed, "base");
	if (tmp1 != nil && tmp1 != [])
	    selected_selection = tmp1[0]:"";

	if (!contains (available_selections, selected_selection))
	    selected_selection = "";

	if (selected_selection != "") {
	    y2milestone ("using installed selection: %1", selected_selection);
	    return;
	}

	// second try: available selection with highest order

	if (size (available_selections) > 0)
	    selected_selection = available_selections[0]:"";

	if (selected_selection != "") {
	    y2milestone ("using highest order selection: %1", selected_selection);
	    return;
	}

	// error: no selection available

	y2error ("no selection available");

    }


    /**
     *
     */
    global define void Detach ()
    {
	Pkg::TargetFinish ();
	did_init1 = false;
	did_init2 = false;
    }

}
