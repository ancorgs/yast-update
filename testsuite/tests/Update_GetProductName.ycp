{

    //testedfiles: Update.ycp

    include "testsuite.ycp";

    map READ = $[
	"target" : $[
	    "size" : 0,
	],
    ];

    list <string> prd_strings = [
	"SUSE LINUX 10.1 (X86-64)\nVERSION = 10.1\n",
	"openSUSE 10.2 (i586) Beta1\nVERSION = 10.2\n",
	"SuSE Linux 9.3 (i586)\nVERSION = 9.3\n",
	"SUSE Linux Enterprise 10 (i586)\nVERSION = 10\n",
	"SUSE LINUX Enterprise Server 9 (ia64)\nVERSION = 9\nPATCHLEVEL = 3",
	"SUSE Linux Enterprise 11 (X86-64) Aplha4\nVERSION = 11\n",
    ];

    map WRITE = $[];

    map EXEC = $[
	"target" : $[
	    "bash_output" : $["exit" : 0, "stdout" : "", "stderr" : ""],
	],
    ];

    TESTSUITE_INIT ([READ, WRITE, EXEC], nil);

    import "Update";
    import "Mode";

    Mode::SetTest("testsuite");

    DUMP ("This test should actually log \"Skipping detection of new system\" && \"true\" because it fails to initialize installation source");

    // this tests only the old system
    // not the new system because it would need to initialize sources
    foreach (string prd_string, prd_strings, {
	READ["target","string"] = prd_string;
	DUMP ("------------------------------------------------------------");
	TEST (``(Update::GetProductName()), [READ, WRITE, EXEC], nil);
	DUMP (prd_string);
	DUMP (Update::installedVersion);
    });

}
