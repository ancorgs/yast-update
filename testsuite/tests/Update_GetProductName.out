Dump	This test should actually log "Skipping detection of new system" && "true" because it fails to initialize installation source
Dump	------------------------------------------------------------
Log	Skipping detection of new system
Return	true
Dump	SUSE LINUX 10.1 (X86-64)
Dump	VERSION = 10.1
Dump	
Dump	$["major":10, "minor":1, "name":"SUSE LINUX", "nameandversion":"SUSE LINUX 10.1", "version":"10.1"]
Dump	------------------------------------------------------------
Log	Skipping detection of new system
Return	true
Dump	openSUSE 10.2 (i586) Beta1
Dump	VERSION = 10.2
Dump	
Dump	$["major":10, "minor":2, "name":"openSUSE", "nameandversion":"openSUSE 10.2", "version":"10.2"]
Dump	------------------------------------------------------------
Log	Skipping detection of new system
Return	true
Dump	SuSE Linux 9.3 (i586)
Dump	VERSION = 9.3
Dump	
Dump	$["major":9, "minor":3, "name":"SuSE Linux", "nameandversion":"SuSE Linux 9.3", "version":"9.3"]
Dump	------------------------------------------------------------
Log	Skipping detection of new system
Return	true
Dump	SUSE Linux Enterprise 10 (i586)
Dump	VERSION = 10
Dump	
Dump	$["major":10, "name":"SUSE Linux Enterprise", "nameandversion":"SUSE Linux Enterprise 10", "version":"10"]
Dump	------------------------------------------------------------
Log	Skipping detection of new system
Return	true
Dump	SUSE LINUX Enterprise Server 9 (ia64)
Dump	VERSION = 9
Dump	PATCHLEVEL = 3
Dump	$["major":9, "name":"SUSE LINUX Enterprise Server", "nameandversion":"SUSE LINUX Enterprise Server 9", "version":"9"]
Dump	------------------------------------------------------------
Log	Skipping detection of new system
Return	true
Dump	SUSE Linux Enterprise 11 (X86-64) Aplha4
Dump	VERSION = 11
Dump	
Dump	$["major":11, "name":"SUSE Linux Enterprise", "nameandversion":"SUSE Linux Enterprise 11", "version":"11"]
